Added check to detect holes in varaible map

amarekano opened the pull request:
This check ensures that we have a compact variable map when importing/generating programs

saelo commented on Sources/Fuzzilli/FuzzIL/Program.swift:
```diff
@@ -193,6 +193,12 @@ public class Program: Collection, Codable {
         }
+
+        // Ensure that the variable map does not contain holes
+        if definedVariables.hasHoles() {
```
Nit: can you change this to `guard !definedVariables.hasHoles() else {` to be consistent with the rest of this function?

amarekano replied:
I've fixed this in the most recent commit.

saelo commented on Sources/Fuzzilli/Util/VariableMap.swift:
```diff
@@ -55,6 +55,10 @@ public struct VariableMap<Value>: Sequence {
     }
+
+    public func hasHoles() -> Bool {
+        return elements.contains(where: {$0 == nil})
```
Could this also just be `elements.contains(nil)` instead?

amarekano replied:
I tried using contains(nil) but that throws a compile error: referencing instance method 'contains' on 'Optional' requires that 'Value' conform to 'Equatable' which is why I resorted to using contains(where:)

saelo replied:
Ah weird, well in that case ignore my comment :)

amarekano commented:
Thanks for the review and I shall add the testcases as soon as I get some time. At the moment, I'm seeing assertion failures when some programs are being minimised. The InlineReducer seems to be generating maps with holes and I'm trying to debug as to why that might be the case. I'm still in the process of familiarising myself with the minimisation phase, would you have thoughts on what some of the potential causes could be? When I get a chance, I'll post some programs (FuzzIL and lifted output) of the misbehaving ones.

saelo commented:
So the minimizer will delete instructions from a program to see whether they are important. This will leave the resulting program with holes in the variable numbers, and so if the Program is checked at that point, the validation would fail. One option might be to just [normalize](https://github.com/googleprojectzero/fuzzilli/blob/master/Sources/Fuzzilli/FuzzIL/Program.swift#L98) the program after every removal done by the minimizer, it should be cheap enough.

amarekano commented:
Thanks @saelo, I've added the new testcases as requested and I've also had to modify one of the existing ones due to the way how shrinkIfNecessary() now works. Adding the normalisation call before ProgramBuilder.finish() returns fixed the issue of holes being generated from program minimisation.
I've also added another comparison operator (unrelated to the varMap checks) but if you'd like that tracked as part of another PR, then I can revert that change.

saelo commented on Sources/Fuzzilli/Core/ProgramBuilder.swift:
```diff
@@ -48,6 +48,7 @@ public class ProgramBuilder {
     ///
     /// The builder instance can not be used further after calling this function.
     public func finish() -> Program {
+        program.normalize()
```
I'm not sure this really fixes the issue. 
I think the main problem with minimization is that it relies on generating programs that are invalid in the new sense (they contain holes in variables), but otherwise correct. As such, this test: https://github.com/googleprojectzero/fuzzilli/blob/5aa6f4ef3724f197a0fc55ffc9b8caf8c809ad56/Sources/Fuzzilli/Minimization/Reducer.swift#L35 should now faill basically all the time, resulting in programs not being minimized. You might see that the average program size keeps increasing or stays very large.
I think you also can't just do `normalize` before that line in the reducer, since some programs will be invalid in other ways, thus probably causing normalize to crash (I think). 
Maybe the correct fix is to add a parameter to the `check` function: `checkForVariableHoles: bool = true` which is true by default but turned off in the Reducer with a comment like `// Due to the way the reducers work, they will produce otherwise valid programs, but with variable holes, so we don't check for those. The holes will be removed after minimization is done`.

amarekano replied:
Thanks for this suggestion, I think this would be the optimal approach and I'll add it to the PR. From what I gather we would need to call `check` with the `checkForVariableHoles` disabled in two places: one in `Reducer.swift` as you highlighted above and the other in `InliningReducer.swift`:

https://github.com/googleprojectzero/fuzzilli/blob/e26e7f78f4b364f5431547f68a412d558e56b07a/Sources/Fuzzilli/Minimization/InliningReducer.swift#L176

amarekano replied:
So disabling the hasHoles check InliningReducer as I mentioned above, turns out, isn't sufficient. Before the assert is called, there's a call to `ProgramBuilder.finish()` which in-turn calls `Program.check()` before it returns.

https://github.com/googleprojectzero/fuzzilli/blob/e26e7f78f4b364f5431547f68a412d558e56b07a/Sources/Fuzzilli/Minimization/InliningReducer.swift#L175-L176 

https://github.com/googleprojectzero/fuzzilli/blob/e26e7f78f4b364f5431547f68a412d558e56b07a/Sources/Fuzzilli/Core/ProgramBuilder.swift#L50-L55

I'm not sure if disabling the hasHoles check in `ProgramBuilder.finish()` would be the right approach here since it's being referenced by several other components. Thoughts?

saelo replied:
Cool, I think the InliningReducer should be fixed now: https://github.com/googleprojectzero/fuzzilli/commit/b8e1778ca24d4bfa8aae5932c6f4a04385cf9958#diff-00251930897a0e17dddc9fdd75f5acf2

With that, does it work now when only setting checkForVariableHoles=false in the ReductionVerifier?

amarekano replied:
Brilliant, I think this would now work. Let me run a few sanity checks and update this PR. I'll try to get this done in the next hour or so :)

amarekano replied:
I've updated the PR to have the reducers ignore holes in the variable Map. I've also reverted the changes to ProgramBuilder since they are no longer necessary with your recent updates to the InliningReducer.

saelo commented on Tests/FuzzilliTests/VariableMapTest.swift:
```diff
@@ -35,7 +35,8 @@ class VariableMapTests: XCTestCase {
         m.remove(v(1))
-        XCTAssert(!m.contains(v(1)) && m[v(1)] == nil)
+
+        XCTAssert(!m.contains(v(42)) && m[v(42)] == nil)
```
This was correct the way it was before, i.e. `XCTAssert(!m.contains(v(1)) && m[v(1)] == nil)`

amarekano replied:
Changes reverted to the original implementation

saelo commented on Tests/FuzzilliTests/VariableMapTest.swift:
```diff
@@ -59,30 +60,33 @@ class VariableMapTests: XCTestCase {
         
         // Add another 128 elements and compare with a new map built up in the opposite order
-        for i in 128..<256 {
+        for i in 127..<255 {
```
Same, should be `128..<256`

amarekano replied:
Changes reverted to the original implementation

saelo commented on Tests/FuzzilliTests/VariableMapTest.swift:
```diff
@@ -59,30 +60,33 @@ class VariableMapTests: XCTestCase {
-        
-        for i in (0..<256).reversed() {
+
+        for i in (0..<255).reversed() {
```
Same, `256`

amarekano replied:
Changes reverted to the original implementation

saelo commented on Tests/FuzzilliTests/VariableMapTest.swift:
```diff
@@ -59,30 +60,33 @@ class VariableMapTests: XCTestCase {
+        for i in (0..<255).reversed() {
             m3[v(i)] = m2[v(i)] ?? false
         }
+
```
Same, you can just revert all the previously existing test cases to how they were before. After all, the class should still behave the same way it did before :)

amarekano replied:
Done, changes to the existing testcases have now been reverted to the original implementation.

saelo commented on Tests/FuzzilliTests/VariableMapTest.swift:
```diff
@@ -135,6 +139,58 @@ class VariableMapTests: XCTestCase {
+        }
+        m.remove(v(0))
+
+        XCTAssertEqual(m.hasHoles(), false)
```
This should now be `true`

amarekano replied:
Testcase updated

saelo commented on Tests/FuzzilliTests/VariableMapTest.swift:
```diff
@@ -135,6 +139,58 @@ class VariableMapTests: XCTestCase {
+        }
+        m.remove(v(Int.random(in: 0..<mapSize)))
+
+        XCTAssertEqual(m.hasHoles(), false)
```
This should now be `true`

amarekano replied:
Testcase updated

amarekano commented on Sources/Fuzzilli/Util/VariableMap.swift:
```diff
@@ -31,11 +31,11 @@ public struct VariableMap<Value>: Sequence {
-        while elements.count > 0 && elements.last! == nil {
-            elements.removeLast()
+        while elements.count > 0 && self.hasHoles() {
+            elements = elements.filter { $0 != nil }
```
Ah I see, I misunderstood the requirement for hasHoles. I assumed, incorrectly, that when a variable is removed, the var map should be compacted to remove holes.
Having re-read the testcases you listed in your first comment to this PR, what I gather is hasHoles should only return true, if there are one or more nil elements at the end of a non empty var map. In all other cases hasHoles should return false.

amarekano replied:
Changes to shrinkIfNecessary have been reverted to the original implementation.

amarekano commented on Sources/Fuzzilli/FuzzIL/Operations.swift:
```diff
@@ -463,6 +463,7 @@ public enum Comparator: String {
     case equal              = "=="
     case strictEqual        = "==="
     case notEqual           = "!="
+    case strictNotEqual     = "!=="
```
No worries, I'll create another PR for this one.

saelo commented:
Merged! Thanks for your work!
